redis/src/aio/connection.rs

```diff
@@ -508,3 +510,101 @@ pub(crate) async fn connect_simple<T: RedisRuntime>(
+/// * `keys` - Array of keys to watch for changes during the transaction
+/// * `func` - Closure that implements the transaction logic. Will receive a cloned connection
+///   and an atomic pipeline. Should return `Result<Option<T>, RedisError>` where:
+///   - `None` indicates the transaction should be retried
+///   - `Some(T)` contains the final result if successful
+///   - `Err` aborts the transaction
+///
+/// # Type Parameters
+///
+/// * `C` - Connection type implementing ConnectionLike + Clone
+/// * `K` - Key type implementing ToRedisArgs
+/// * `T` - Return value type implementing FromRedisValue
+/// * `F` - Closure type taking (C, Pipeline) and returning Fut
+/// * `Fut` - Future returning Result<Option<T>, RedisError>
+///
+/// # Examples
+///
+/// ```rust,no_run
+/// use redis::{AsyncCommands, RedisResult, pipe};
+///
+/// async fn increment(con: redis::aio::ConnectionManager) -> RedisResult<isize> {
+///     let key = "my_counter";
+///     redis::aio::transaction_async(con, &[key], |mut con, mut pipe| async move {
+///         //  Get the current value
+///         let val: isize = con.get(key).await?;
+///         //   Create pipeline that increments and returns new value
+///         pipe.set(key, val + 1)
+///             .ignore()
+///             .get(key)
+///             .query_async(&mut con)
+///             .await
+///     })
+///     .await
+/// }
+/// ```
+///
+/// # Notes
+///
+/// - The closure provided to this function may be executed multiple times if the
+///   watched keys are modified by other clients between the WATCH and EXEC commands
+/// - Side effects in the closure should be idempotent since they may occur multiple times
+/// - The transaction is automatically rolled back if the closure returns an error
+/// - Returns the final result once the transaction succeeds
+///
+/// For more details on Redis transactions, see the [Redis documentation](https://redis.io/topics/transactions)
+pub async fn transaction_async<
+    Connection: ConnectionLike + Clone,
+    K: ToRedisArgs,
+    T: FromRedisValue,
+>(
+    mut connection: Connection,
+    keys: &[K],
+    pipeline: &mut Pipeline,
+) -> Result<T, RedisError> {
+    let mut transaction_pipeline = pipe(); // a long lived value that will be used to build the transaction pipeline
+    transaction_pipeline.cmd("WATCH").arg(keys).cmd("MULTI");
+
+    std::mem::take(&mut pipeline.commands) // consume the pipeline and give into_iter() ownership of the commands attribute
+        .into_iter()
+        .filter(|cmd| cmd.data != b"UNWATCH")
+        .for_each(|cmd| {
+            transaction_pipeline.add_command(cmd);
+        });
+    transaction_pipeline.cmd("EXEC");
+
+    loop {
+        let response: Result<Option<Vec<_>>, RedisError> =
+            transaction_pipeline.query_async(&mut connection).await;
+        cmd("UNWATCH").exec_async(&mut connection).await?; //  // Always clean up the WATCH after the transaction attempt, regardless of success/failure
```

Comment:
We may put the `UNWATCH` command directly at the very end of the pipeline ?

Reply:
I don't think that the UNWATCH won't be executed. a `Pipeline` object is a collection of commands sent to the server. The server sees them just as if each command on the pipeline was sent separately, and early commands failure doesn't affect the next commands just because they're in a pipeline together. so executing command after a pipeline is equivalent to adding it to the pipeline - only, if the command is added to the pipeline, no other command can be interleaved, so its actually safer.